### PR TITLE
Feature/data object

### PR DESCRIPTION
**Pull request by Penfu**
@NicolasGlassey Vous pouvez review cette PR.

- Classe de test 'TestGCPDataObject' incomplète.
- Github actions

### CODE REVIEW

**Review by MauroWasTaken — APPROVED** [review]
the application side of this PR will be reviewed in seperate features

File: doc/sequences/publish.puml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        Mauro COSTA-DOS-SANTOS 
+endlegend
+
+actor User
+
+participant "DataObject" as App
+participant "GCP" as GCP
+
+User -> App: PublishObject
+App -> GCP: DoesExist Request
```
Comment (NicolasGlassey) [inline doc/sequences/publish.puml:21]: Is it really the responsibility of the Publish Method to test if the object does exist?

The SDK already returns an exception if the object wasn't found. Publish has to correctly deal with this exception, but for me, the client (the class that calls the publish method) is responsible of this.

File: doc/sequences/download.puml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        Mauro COSTA-DOS-SANTOS 
+endlegend
+
+actor User
+
+participant "DataObject" as App
+participant "GCP" as GCP
+
+User -> App: DownloadObject
+App -> GCP: DoesExist Request
```
Comment (NicolasGlassey) [inline doc/sequences/download.puml:21]: Same issue... the same question of responsibility. The client is responsible to use correctly the create method. Are any other exceptions you have to deal with?

**Review by NicolasGlassey — CHANGES_REQUESTED** [review]
@MauroWasTaken  @Penfu 

Please find as comments, some issues I had on several .puml files (sequences).

Globally, be careful of the concern of responsibility. The "DoesExist" method is already responsible to check if an object exists. All other methods are specialized for other purposes. This is the client who is responsible to use correctly the method (pre-conditions).

I propose you merge those singular sequence diagrams in a real scenario. The sequence must be more "integration test" oriented.

Publish a new object from its creation, with all possible cases. I request only one sequence, but with everything, we need for the DataObjectHelper.

File: doc/sequences/exist.puml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+actor User
+
+participant "DataObject" as App
+participant "GCP" as GCP
+
+User -> App: DoesExist
+App -> GCP: GetObject Request
+GCP -> App: GetObject Response
+
```
[the remark this answers is not in the document]
Comment (Penfu) [reply doc/sequences/exist.puml:23]: @NicolasGlassey We didn't found an SDK-level method to check if an object exists. 

That's why we perform a get of the object to the SDK  and he return the object or an exception.
Base on this answer we return to the client the corresponding bolean statement. The SDK exception isn't thrown further.

File: doc/sequences/publish.puml (hunk 1 above)
Comment (Penfu) [reply doc/sequences/publish.puml]: Solve in this [commit](https://github.com/CPNV-ES-BI/BI_CSHARP_GCP/commit/6ca4b6d6232df5449d33513c6bcd197fac593079).

File: doc/sequences/download.puml (hunk 2 above)
Comment (Penfu) [reply doc/sequences/download.puml]: Solve in this [commit](https://github.com/CPNV-ES-BI/BI_CSHARP_GCP/commit/5f67e667425ec220665b3b1d474be7c7bf44952b).

It is also possible to receive an exception from the SDK regarding the client's access rights to the object. But we will not handle this case in this sprint.

**Comment by Penfu** [comment]
# Publish

## Issue

Google.GoogleApiException : 'The service storage has thrown an exception. HttpStatusCode is BadRequest. Cannot update access control for an object when uniform bucket-level access is enabled. Read more at https://cloud.google.com/storage/docs/uniform-bucket-level-access'

## Solution

To allow to attach specific access  right to an object change bucket policy with: `gsutil uniformbucketlevelaccess set off gs://[BUCKET_NAME]`

**Comment by Penfu** [comment]
# Publish public access prevention is enforced

## Issue

Google.GoogleApiException : 'The service storage has thrown an exception. HttpStatusCode is PreconditionFailed. The member bindings allUsers and allAuthenticatedUsers are not allowed since public access prevention is enforced.'

## Solution

Go to Storage browser in the GCP Console, find your bucket, click on the three dots on the right, and select Edit bucket then click on Permissions tab and uncheck Block all public access.

Didn't find the CLI.

**Comment by Penfu** [comment]
In order to implement the API, this PR is merge.

**Pull request merged**